src/posthog-featureflags.ts

```diff
@@ -483,4 +484,118 @@ export class PostHogFeatureFlags {
             this.instance.unregister(STORED_GROUP_PROPERTIES_KEY)
         }
     }
+
+    _getMatchingVariant(featureFlag: ClientAssignedFeatureFlag): string | null {
+        const lookupTable = this._variantLookupTable(featureFlag.variants)
+        const hash = this._get_hash(featureFlag.key, this.instance.get_distinct_id(), 'variant')
+
+        for (const variant of lookupTable) {
+            if (hash >= variant.value_min && hash < variant.value_max) {
+                return variant.key
+            }
+        }
+        return null
+    }
+
+    // TODO how should this behave for erroneous values?
+    _variantLookupTable(variants: Record<string, number>): { value_min: number; value_max: number; key: string }[] {
+        const lookupTable: { value_min: number; value_max: number; key: string }[] = []
+        let valueMin = 0
+
+        for (const [variant, percentage] of Object.entries(variants)) {
+            const valueMax = valueMin + percentage
+            lookupTable.push({
+                value_min: valueMin,
+                value_max: valueMax,
+                key: variant,
+            })
+            valueMin = valueMax
+        }
+        return lookupTable
+    }
+
+    _get_hash(featureFlagKey: string, distinctId: string, salt: string = ''): number {
+        const hashKey = `${featureFlagKey}.${distinctId}${salt}`
+        const hashHex = this._hash(hashKey)
+        // TODO do we care about IE11 support for BigInt?
+        const hashInt = BigInt(`0x${hashHex}`)
+        const LONG_SCALE = BigInt('0xFFFFFFFFFFFFFFF')
+        return Number(hashInt) / Number(LONG_SCALE) // Normalize the hash to a value between 0 and 1
+    }
+
+    // TODO how much do we trust sonnet to write a hashing function?
+    _hash(input: string): string {
```

Comment:
we use `sha1` in both the python and rust implementations when creating hash keys; any reason why we wouldn't do the same here?

Reply:
> any reason why we wouldn't do the same here?

@dmarticus Tell me more? I ran into some IE11 linting issue with `crypto`, but [maybe we can actually use it](https://caniuse.com/cryptography).

Reply:
argh, should've known there's some browser compat issue – there's some prior art ([slack](https://posthog.slack.com/archives/C0113360FFV/p1729152330844419), [github](https://github.com/PostHog/posthog-js/pull/1480) around separating our bundles into es5 (for things that explicitly need IE11 support) and es6 (for everything else), but I'm not sure what our policy on adding new behavior that explicitly doesn't support IE11 is.   

It might be a pain to support it here though – the decision is up to you, I just wanted to provide my 2 cents.  I definitely didn't realize this was a browser compat thing beforehand.